// ibmsecurity/js/lib/util/constants.js
/*!
 *  Licensed Materials - Property of IBM
 *  5725-I43 (C) Copyright IBM Corp. 2011, 2014. All Rights Reserved.
 *  US Government Users Restricted Rights - Use, duplication or
 *  disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 *
 *  IBM Mobile Application Security (MAS) v0.0.1
 */

'use strict';

var Constants = {
		MAS_NAME        : 'mas',
        MAS_VERSION     : 'v1',
        MAS_URI_APPS    : 'apps',
        MAS_PREFIX      : 'MAS',
        MAS_REQUIRESSL_OPTION  : 'requiressl', // case is important
        MAMUI_WEB_CONTEXT_NAME : 'masui',
        MAMSERVICE_INTERNAL_SUBDOMAIN_NAME : "MAS",

        HEADERS    : { // Older header mappings for v0.3.8 to be deprecated
            "access_token"      : "IBM-Security-Token",
            "oauth_provider"    : "IBM-Security-Provider",
            "platform"          : ["Client-Platform-Type", "IBM-Device-Type"],
            "package"           : ["Client-Id","IBM-Package-ID"],
            "appId"             : "IBM-Application-ID",
            "dev_access_token"  : "Access-Token",
            "devId"             : "IBM-Device-ID",
            "devmodel"          : "ibm-device-model", 
            "devversion"        : "ibm-device-platform-version", 
            "devtype"           : "ibm-device-type"                
        },
        EVENTTYPES : {
            "MISSING_ACCESS_TOKEN"          : "MISSING_ACCESS_TOKEN",
            "MISSING_OAUTH_PROVIDER"        : "MISSING_OAUTH_PROVIDER",
            "INVALID_OAUTH_PROVIDER_CONFIG" : "INVALID_OAUTH_PROVIDER_CONFIG",
            "INVALID_OAUTH_TOKEN"           : "INVALID_OAUTH_TOKEN",
            "INVALID_CLIENT_REQUEST"        : "INVALID_CLIENT_REQUEST",
            "CLIENT_NOT_ENABLED"            : "CLIENT_NOT_ENABLED",
            "INVALID_APP_CONFIG"            : "INVALID_APP_CONFIG",
            "UNDEFINED"                     : "UNDEFINED"
        },
        HTTP_RESP_CODES : {
            "Success" : 200,
            "Create" : 201,
            "BadRequest" : 400,
            "NotAuthorized" : 401,
            "NotFound" : 404,
            "Conflict" : 409,
            "InternalError" : 500
        },
        HTTP_REQUEST_HEADERS : {
            "access_token"      : "ibm-security-token",
            "oauth_provider"    : "ibm-security-provider",
            "package"           : "ibm-package-id",
            "appSecret"         : "ibm-application-secret",
            "appversion"        : "ibm-application-version",
            "dev_access_token"  : "access-token",
            "devId"             : "ibm-device-id",
            "devlocation"       : "ibm-device-location",
            "devmodel"          : "ibm-device-model",
            "devversion"        : "ibm-device-platform-version",
            "devtype"           : "ibm-device-type",
            "devname"           : "ibm-device-name",
            "accept_lang"       : "accept-language",
            "sdk_core_ver"      : "ibm-sdk-core-version",
            "sdk_service_ver"   : "ibm-sdk-service-version",
            "sdk_service_name"  : "ibm-sdk-service-name"
        }
}

exports = module.exports = Constants;
